statev2/src/storage/db.rs

```diff
@@ -0,0 +1,207 @@
+//! Defines the interface to the `mdbx` instance
+//!
+//! We serialize types using the `flexbuffers` format (a schema-less version of
+//! `flatbuffers`): https://flatbuffers.dev/flexbuffers.html
+
+use std::{borrow::Cow, path::Path};
+
+use libmdbx::{Database, TableFlags, Transaction, WriteFlags, WriteMap, RO, RW};
+use serde::{Deserialize, Serialize};
+
+use super::{
+    error::StorageError,
+    traits::{Key, Value},
+};
+
+/// The number of tables to open in the database
+const NUM_TABLES: usize = 2;
+
+/// The database config
+pub struct DbConfig {
+    /// The path to open the database at
+    path: String,
+}
+
+/// The persistent storage layer for the relayer's state machine
+///
+/// Contains a reference to an `mdbx` instance
+pub struct DB {
+    /// The underlying `mdbx` instance
+    db: Database<WriteMap>,
+    /// The config for the database
+    #[allow(unused)]
+    config: DbConfig,
+}
+
+impl DB {
+    /// Constructor
+    pub fn new(config: DbConfig) -> Result<Self, StorageError> {
+        let db_path = Path::new(&config.path);
+        let db = Database::new()
+            .set_max_tables(NUM_TABLES)
+            .open(db_path)
+            .map_err(StorageError::OpenDb)?;
+
+        Ok(Self { db, config })
+    }
+
+    /// Create a new table in the database
+    pub fn create_table(&self, table_name: &str) -> Result<(), StorageError> {
+        // Begin a tx
+        let tx = self.new_write_tx()?;
+        tx.create_table(Some(table_name), TableFlags::default())
+            .map_err(StorageError::TxOp)
+            .map(|_| ())
+            .and_then(|_| tx.commit().map_err(StorageError::Commit).map(|_| ()))
+    }
+
+    /// Set a key in the database
+    pub fn set_table<K: Key, V: Value>(
+        &self,
+        table_name: &str,
+        key: &K,
+        value: &V,
+    ) -> Result<(), StorageError> {
+        // Serialize the key and value
+        let key_bytes = Self::serialize_value(key)?;
+        let value_bytes = Self::serialize_value(value)?;
+
+        // Begin a tx
+        let tx = self.new_write_tx()?;
+        let table = tx
+            .open_table(Some(table_name))
+            .map_err(StorageError::OpenTable)?;
+
+        // Set the value
+        tx.put(&table, key_bytes, value_bytes, WriteFlags::default())
+            .map_err(StorageError::TxOp)
+            .and_then(|_| tx.commit().map_err(StorageError::Commit).map(|_| ()))
+    }
+
+    /// Get a key from the database
+    pub fn get_table<K: Key, V: Value>(
```

Comment:
nit: `get_table` is a slightly confusing name, my initial reaction would be that this returns an ORM-style handle on a table rather than look up a single value in a table (same for `set_table` above).

i think the function signature documents itself pretty well with the `table_name` parameter, but if just `get` and `set` feel too vague, maybe we could do `get_value` and `set_value`?